src/database/models/image_item.rs

```diff
@@ -0,0 +1,329 @@
+use super::ids::*;
+use crate::database::models::DatabaseError;
+use crate::models::ids::base62_impl::{parse_base62, to_base62};
+use chrono::{DateTime, Utc};
+use redis::cmd;
+use serde::{Deserialize, Serialize};
+
+const IMAGES_NAMESPACE: &str = "images";
+const IMAGE_URLS_NAMESPACE: &str = "image_urls";
+const DEFAULT_EXPIRY: i64 = 1800; // 30 minutes
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Image {
+    pub id: ImageId,
+    pub url: String,
+    pub size: u64,
+    pub created: DateTime<Utc>,
+    pub owner_id: UserId,
+
+    pub mod_id: Option<ProjectId>,
+    pub thread_message_id: Option<ThreadMessageId>,
+}
+
+impl Image {
+    pub async fn insert(
+        &self,
+        transaction: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    ) -> Result<(), DatabaseError> {
+        sqlx::query!(
+            "
+            INSERT INTO uploaded_images (
+                id, url, size, created, owner_id, mod_id, thread_message_id
+            )
+            VALUES (
+                $1, $2, $3, $4, $5, $6, $7
+            );
+            ",
+            self.id as ImageId,
+            self.url,
+            self.size as i64,
+            self.created,
+            self.owner_id as UserId,
+            self.mod_id.map(|x| x.0),
+            self.thread_message_id.map(|x| x.0),
+        )
+        .execute(&mut *transaction)
+        .await?;
+
+        Ok(())
+    }
+
+    pub async fn remove(
+        id: ImageId,
+        transaction: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+        redis: &deadpool_redis::Pool,
+    ) -> Result<Option<()>, DatabaseError> {
+        let image = Self::get_id(id, &mut *transaction, redis).await?;
+
+        if let Some(image) = image {
+            sqlx::query!(
+                "
+                DELETE FROM uploaded_images
+                WHERE id = $1
+                ",
+                id as ImageId,
+            )
+            .execute(&mut *transaction)
+            .await?;
+
+            Image::clear_cache(image.id, image.url, redis).await?;
+
+            Ok(Some(()))
+        } else {
+            Ok(None)
+        }
+    }
+
```

Comment:
when would all of these functions get used?

Reply:
When we want to insert or delete images due to the relevant thing being inserted/deleted- so the deletion one is called if a relevant project or thread message is deleted for instance